### PR TITLE
R gotcha blog

### PR DESCRIPTION
**Pull request by weshinsley**
A blog about two ways I have become unstuck in programming R...
The usual sort of "flustered and doomed" narrative that is common to my blogs so far!

I also realised during writing that I'm not sure what Rob and James's favourite food is.... so please pay particular attention to that important detail in the review... (Alex, Rich, Emma, I'm pretty confident I've got right...)

### CODE REVIEW

File: content/blog/gotchaR-bumps.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+Great. However, I later want to do something astoundingly complicated with this rich dataset, and I 
+decide that using the popular [data.table](https://rdatatable.gitlab.io/data.table/) will 
+be a good idea, since it's just a higher performance version of `data.frame`, right? It's
+also in very common use. (To be clear, the performance increase is indeed very worthwhile for large
+datasets).
+
+However... if I swap in the `data.table` by only changing 
+```
+data <- data.frame(stringsAsFactors = FALSE
```
Comment (richfitz) [inline content/blog/gotchaR-bumps.md]: missing closing paren here and line 113 - consider using `...` as `data.frame(..., stringsAsFactors = FALSE)` perhaps?

File: content/blog/gotchaR-bumps.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+"Fruit". It doesn't auto-complete on the assignment side though, so after the first
+loop iteration, `data$thing` exists, and things continue normally. 
+
+It's tempting to think
+`data$thing <- NULL` beforehand would help - but with a list, that
+would only remove the (non-existent) `thing` from list, because that's what NULL assignment
+with lists does.
+
+So what should I instead do? One option is to replace my inner loop with
+` data$thing <- c(data[['thing']], f)`, because the auto-completion is activated on the
```
Comment (richfitz) [inline content/blog/gotchaR-bumps.md]: stray space after the opening backtick

File: content/blog/gotchaR-bumps.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+All useless. I start to wonder if R has been tricking me all along, and I do some really
+basic assignments and concatenations to check it's not been fatally flawed all these years. 
+It appears fundamentally sound. 
+
+Eventually, I ask my more R-experienced colleagues if, in fact, I am going insane. They
+cannot answer this for sure, but looking at the programming issue, they tell me it is a 
+feature of R in a way. 
+
+The first time I read from `data$thing`, it doesn't exist 
+in the list, and R lexically auto-completes it into `data$thing_type` - hence the
```
Comment (richfitz) [inline content/blog/gotchaR-bumps.md]: I would avoid "lexically" here as it is usually used to refer to [lexical scope](https://en.wikipedia.org/wiki/Scope_(computer_science)#Lexical_scoping) whereas this is just autocompletion of a sort

File: content/blog/gotchaR-bumps.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+with lists does.
+
+So what should I instead do? One option is to replace my inner loop with
+` data$thing <- c(data[['thing']], f)`, because the auto-completion is activated on the 
+use of the dollar sign, but not on a quoted variable name. Moreover, setting these options:
+```
+warnPartialMatchAttr = TRUE
+warnPartialMatchDollar = TRUE
+warnPartialMatchArgs = TRUE
+``` in `.Rprofile` will throw a warning every time R does a partial-match. This seems like
```
Comment (richfitz) [inline content/blog/gotchaR-bumps.md]: newline after the closing triple-backtick to avoid formatting issues

File: content/blog/gotchaR-bumps.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+it might save a world of pain, although casting my mind back to various projects where I have
+frivolously talked about `list$thing` and `list$thing_name` and `list$thing_id` still makes me
+shudder a little. 
+
+Also having set these options, I occasionally see warnings popping up about auto-completion
+in code for other libraries we rely upon...
```
Comment (richfitz) [inline content/blog/gotchaR-bumps.md]: "libraries" should be "packages"

File: content/blog/gotchaR-bumps.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+}
+```
+
+then all works as I expected. And if I (oddly) changed `x` to `name`, then data.table
+would perform comparison between columns, returning zero rows.
+
+Looking through my previous code I have quite a number of examples where the most natural
+thing to write would be `data[data$country == country, ]` or `data[data$id == id, ]`. Luckily
+I did so only with data-frames. I have not yet found any warnings I can switch on,
+should I forget about this issue in the future. Again.
```
Comment (richfitz) [inline content/blog/gotchaR-bumps.md:142]: I had a look to see what happens with tibbles here and got:

```
5 Rob       Celery 
> data <- tibble::tibble(data)
> who_likes <- function(food) {
+   data[data$food == food, ]
+ }
> who_likes("Goose")
# A tibble: 0 x 1
# … with 2 variables: data$name <chr>, $food <chr>
Warning messages:
1: Unknown or uninitialised column: 'food'. 
2: Length of logical index must be 1 or 5, not 0 
```

which seems somewhat reasonable

File: content/blog/gotchaR-bumps.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+and some dabblings with assemblers of various breeds. Mostly, my adjustment has been
+about doing things the way R likes; turning loop-ish code into vector-ish code, using
+the right base functions properly and thinking about data in columns for a start.
+
+Occasionally though, R has the capacity to make me doubt either its sanity, or
+my own, at least for a while. Here are a couple of my lowest points. More may come later, but 
+two is a start, and those PRs for [Hacktoberfest](https://hacktoberfest.digitalocean.com/) 
+aren't going to happen on their own...
+
+# The dodgy dollar?
```
Comment (richfitz) [inline content/blog/gotchaR-bumps.md:19]: [dollar dollar bill y'all](https://www.youtube.com/watch?v=PBwAxmrE194)

**Review by EmmaLRussell — APPROVED** [review]
This is great. R is so weird...
Also, if you were to offer me haggis I'd tell you exactly what you could do with it, but that's irrelevant for exemplary purposes..

**Comment by weshinsley** [comment]
> Also, if you were to offer me haggis I'd tell you exactly what you could do with it, but that's irrelevant for exemplary purposes..

@EmmaLRussell I have updated the blog post with a footnote to communicate this! :-)

File: content/blog/gotchaR-bumps.md (hunk 2 above)
Comment (weshinsley) [reply content/blog/gotchaR-bumps.md]: Done

File: content/blog/gotchaR-bumps.md (hunk 1 above)
Comment (weshinsley) [reply content/blog/gotchaR-bumps.md]: I've cleaned that up a bit into a tighter sentence - it was all a bit spaced out anyway.

File: content/blog/gotchaR-bumps.md (hunk 3 above)
Comment (weshinsley) [reply content/blog/gotchaR-bumps.md]: Done.

File: content/blog/gotchaR-bumps.md (hunk 4 above)
Comment (weshinsley) [reply content/blog/gotchaR-bumps.md]: Done

File: content/blog/gotchaR-bumps.md (hunk 5 above)
Comment (weshinsley) [reply content/blog/gotchaR-bumps.md]: Done

File: content/blog/gotchaR-bumps.md (hunk 7 above)
Comment (weshinsley) [reply content/blog/gotchaR-bumps.md:19]: Amazing...

**Review by richfitz — APPROVED** [review]

**Pull request merged**